pkg/runner/executable_path_resolving_runner.go

```diff
@@ -0,0 +1,121 @@
+package runner
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type rootDirectoryOpener func(inputRootDirectory string) (filesystem.DirectoryCloser, error)
+
+type executablePathResolvingRunner struct {
+	base                Runner
+	rootDirectoryOpener rootDirectoryOpener
+}
+
+// NewExecutablePathResolvingRunner modifies the arguments of the
+// incoming RunRequest so that the first argument is an absolute path
+// resolved using the path in the request, using the input root as the
+// chroot if necessary
+func NewExecutablePathResolvingRunner(base Runner, rootDirectoryOpener rootDirectoryOpener) Runner {
+	return &executablePathResolvingRunner{
+		base:                base,
+		rootDirectoryOpener: rootDirectoryOpener,
+	}
+}
+
+func (r *executablePathResolvingRunner) enterDir(file, chroot string) (filesystem.DirectoryCloser, error) {
+	rootDir, err := r.rootDirectoryOpener(chroot)
+	if err != nil {
+		return nil, util.StatusWrap(err, "Failed to open root directory")
+	}
+	components := strings.FieldsFunc(file, func(r rune) bool { return r == '/' })
+	dir := rootDir
+	for n, component := range components[:len(components)-1] {
+		dir2, err := dir.EnterDirectory(component)
```

Comment:
I think this doesn't work when a directory is a symlink, right? For example, consider the case where:
```
PATH=/bin
```
And `/bin` is a symlink to `usr/bin`.

Reply:
Would it be acceptable in the case you described to have the eventual value of `Arguments[0]` still be `/bin/<executable>`. A very quick test suggests that `exec.Command` will still work but there may be an edge case I'm not aware of.